# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources (README.md file)
- Happy learning!

In [1]:
# import numpy and pandas
import pandas as pd
import numpy as np
from scipy.stats import trim_mean, mode, skew, gaussian_kde, pearsonr, spearmanr, beta
from statsmodels.stats.weightstats import ztest as ztest

from scipy.stats import ttest_ind, norm, t
from scipy.stats import f_oneway
from scipy.stats import sem

# Challenge 1 - Exploring the Data

In this challenge, we will examine all salaries of employees of the City of Chicago. We will start by loading the dataset and examining its contents

In [2]:
# Run this code:
salaries = pd.read_csv(r'C:\Users\SAMSUNG\Desktop\DSML_Lab\Week4\Day3\lab-hypothesis-testing-en\data\Current_Employee_Names__Salaries__and_Position_Titles.csv')

Examine the `salaries` dataset using the `head` function below.

In [3]:
salaries.head(5)

,Name,Job Titles,Department,Full or Part-Time,Salary or Hourly,Typical Hours,Annual Salary,Hourly Rate
0,"AARON, JEFFERY M",SERGEANT,POLICE,F,Salary,NaN,101442.0,NaN
1,"AARON, KARINA",POLICE OFFICER (ASSIGNED AS DETECTIVE),POLICE,F,Salary,NaN,94122.0,NaN
2,"AARON, KIMBERLEI R",CHIEF CONTRACT EXPEDITER,GENERAL SERVICES,F,Salary,NaN,101592.0,NaN
3,"ABAD JR, VICENTE M",CIVIL ENGINEER IV,WATER MGMNT,F,Salary,NaN,110064.0,NaN
4,"ABASCAL, REECE E",TRAFFIC CONTROL AIDE-HOURLY,OEMC,P,Hourly,20.0,NaN,19.86


# Challenge 2 - Hypothesis Tests

In this section of the lab, we will test whether the hourly wage of all hourly workers is significantly different from $30/hr. Import the correct one sample test function from scipy and perform the hypothesis test for a 95% two sided confidence interval.

In [4]:
import scipy.stats as st

In [9]:
salaries.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 33183 entries, 0 to 33182
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Name               33183 non-null  object 
 1   Job Titles         33183 non-null  object 
 2   Department         33183 non-null  object 
 3   Full or Part-Time  33183 non-null  object 
 4   Salary or Hourly   33183 non-null  object 
 5   Typical Hours      8022 non-null   float64
 6   Annual Salary      25161 non-null  float64
 7   Hourly Rate        8022 non-null   float64
dtypes: float64(3), object(5)
memory usage: 2.0+ MB


In [5]:
s1=salaries[salaries['Salary or Hourly']=='Hourly']

In [17]:
s1.info()

<class 'pandas.core.frame.DataFrame'>
Index: 8022 entries, 4 to 33175
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Name               8022 non-null   object 
 1   Job Titles         8022 non-null   object 
 2   Department         8022 non-null   object 
 3   Full or Part-Time  8022 non-null   object 
 4   Salary or Hourly   8022 non-null   object 
 5   Typical Hours      8022 non-null   float64
 6   Annual Salary      0 non-null      float64
 7   Hourly Rate        8022 non-null   float64
dtypes: float64(3), object(5)
memory usage: 564.0+ KB


In [12]:
s2=s1['Hourly Rate']

In [18]:
s3=s2.sample(8000)
st.ttest_1samp(s3, 30)

TtestResult(statistic=np.float64(20.581234710571923), pvalue=np.float64(9.344681547548505e-92), df=np.int64(7999))

In [19]:
st.ttest_1samp(s3, 30, alternative="greater")

TtestResult(statistic=np.float64(20.581234710571923), pvalue=np.float64(4.6723407737742526e-92), df=np.int64(7999))

In [20]:
st.ttest_1samp(s3, 30, alternative="less")

TtestResult(statistic=np.float64(20.581234710571923), pvalue=np.float64(1.0), df=np.int64(7999))

In [21]:
s3.describe()

count    8000.000000
mean       32.788195
std        12.117045
min         2.650000
25%        21.200000
50%        35.600000
75%        40.200000
max       109.000000
Name: Hourly Rate, dtype: float64

# Challenge 3 - Constructing Confidence Intervals

While testing our hypothesis is a great way to gather empirical evidence for accepting or rejecting the hypothesis, another way to gather evidence is by creating a confidence interval. A confidence interval gives us information about the true mean of the population. So for a 95% confidence interval, we are 95% sure that the mean of the population is within the confidence interval. 
).

To read more about confidence intervals, click [here](https://en.wikipedia.org/wiki/Confidence_interval).


In the cell below, we will construct a 95% confidence interval for the mean hourly wage of all hourly workers. 

The confidence interval is computed in SciPy using the `t.interval` function. You can read more about this function [here](https://docs.scipy.org/doc/scipy-0.14.0/reference/generated/scipy.stats.t.html).

To compute the confidence interval of the hourly wage, use the 0.95 for the confidence level, number of rows - 1 for degrees of freedom, the mean of the sample for the location parameter and the standard error for the scale. The standard error can be computed using [this](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.sem.html) function in SciPy.

In [22]:
#using numpy and scipy
#1. np.mean() computes the sample mean of data
#2. np.std(d, ddof=1) calculates the sample standard deviation, using ddof=1 for sample std
#3. len(d) determines the number of data points (sample size)
#4. st.t.interval() calculates the confidence interval using the confidence level, degress of freedom, sample mean, and standard error (sample std/square root of sample size)
ci=st.t.interval(0.95, df=len(s3)-1, loc=np.mean(s3), scale=np.std(s3, ddof=1)/np.sqrt(len(s3)))
ci

(np.float64(32.522633245350484), np.float64(33.053756754649534))

# Challenge 4 - Hypothesis Tests of Proportions

Another type of one sample test is a hypothesis test of proportions. In this test, we examine whether the proportion of a group in our sample is significantly different than a fraction. 

You can read more about one sample proportion tests [here](http://sphweb.bumc.bu.edu/otlt/MPH-Modules/BS/SAS/SAS6-CategoricalData/SAS6-CategoricalData2.html).

In the cell below, use the `proportions_ztest` function from `statsmodels` to perform a hypothesis test that will determine whether the number of hourly workers in the City of Chicago is significantly different from 25% at the 95% confidence level.

In [24]:
salaries.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 33183 entries, 0 to 33182
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Name               33183 non-null  object 
 1   Job Titles         33183 non-null  object 
 2   Department         33183 non-null  object 
 3   Full or Part-Time  33183 non-null  object 
 4   Salary or Hourly   33183 non-null  object 
 5   Typical Hours      8022 non-null   float64
 6   Annual Salary      25161 non-null  float64
 7   Hourly Rate        8022 non-null   float64
dtypes: float64(3), object(5)
memory usage: 2.0+ MB


In [25]:
totalworker=salaries['Name'].count()

In [26]:
hourlyworker=salaries['Hourly Rate'].count()

In [ ]:
#count: number of hourly workers in Chicago
#nobs: total number of observations 
#values: hypothesized proportion 


In [27]:
from statsmodels.stats.proportion import proportions_ztest

In [28]:
proportions_ztest(hourlyworker, totalworker, 0.25)


(np.float64(-3.5099964213703005), np.float64(0.0004481127249057967))

Z-statistic is -3.51 and p-value is 0.00045
at 95% confidence level, p<0.05 will reject the null hypothesis
thus, the proportion of hourly workers in Chicago is significantly different from 25 %